_posts/2024-04-23-kubernetes-horizontal-pod-autoscaler-guide.md

```diff
@@ -0,0 +1,164 @@
+{: refdef}
+
+The memory usage seems to be below the 90% so the number of replicas would be set to 1. Let's check the CPU usage now:
+
+<div class="post-center-image">
+    {% picture pimage /assets/images/fullsize/posts/2024-04-23-kubernetes-horizontal-pod-autoscaler-guide/cpu_init.png --alt HPA CPU expected replicas %}
+</div>
+
+{:refdef: class="image-caption"}
+*Figure 3. HPA CPU expected replica*
+{: refdef}
+
+
+At startup, the CPU usage exceeded the target value of `450m`. This means that the 'currentMetricValue / desiredMetricValue' ratio was greater than one, indicating that the autoscaler needed to scale up the replicas. But by how much? Let’s adjust the x-axis of the plot to display the number of replicas:
+
+<div class="post-center-image">
+    {% picture pimage /assets/images/fullsize/posts/2024-04-23-kubernetes-horizontal-pod-autoscaler-guide/cpu_stairs.png --alt HPA CPU expected replicas showing stairs pattern %}
+</div>
+
+{:refdef: class="image-caption"}
+*Figure 4. HPA CPU current replicas vs expected replicas*
+{: refdef}
+
+There you go! We can clearly see a pattern resembling infinite stairs, similar to $f(x) = x + 1$. This occurs because the 'currentMetricValue / desiredMetricValue' ratio is greater than 1, prompting the autoscaler to continuously increase the number of replicas until it reaches the maximum allowed. In this instance, maxReplicas was set to 3. Thus, we have identified the root of the problem!
+
+## Why is HPA not scaling down?
+
+Although CPU usage spiked at startup, it quickly returned to low levels. So why isn't the HPA scaling down the number of replicas? It appears that the CPU requirement is well below the target value of `450m`, as illustrated in Figure 1. However, the bottleneck now is memory usage.
+
+As shown by Grafana, memory usage remains constant. Figure 2 indicated that the expected number of replicas should be 1. However, since the HPA previously scaled up our replicas to 3, visualizing the same plot with the number of replicas on the x-axis reveals the following:
+
+<div class="post-center-image">
+    {% picture pimage /assets/images/fullsize/posts/2024-04-23-kubernetes-horizontal-pod-autoscaler-guide/memory_stairs.png --alt HPA memory current replicas vs expected replicas %}
+</div>
+
+{:refdef: class="image-caption"}
+*Figure 5. HPA memory current replicas vs expected replicas*
+{: refdef}
+
+With the current memory usage, the HPA behaves like the function $f(x) = x$, preventing the number of replicas from scaling down. This is why we continuously see three pods running, even though the pods are not receiving much traffic.
+
+## Conclusion
+
+In this post, we looked at how the Horizontal Pod Autoscaler (HPA) in Kubernetes helps manage the number of pods based on CPU and memory use. This is very important for keeping applications running smoothly as demands change. However, we also learned that it's crucial to keep an eye on and adjust the HPA settings in real-world use.
+
```

Comment:
Maybe you'll disagree here, we can discuss it if that's the case. But for me, the key takeaway is that you should scale based on metrics dominated by variable load, not fixed load. And I would emphasize this in the conclusions, maybe showing the figure I include below to illustrate it.

I think the underlying reason we run into this problem is memory usage was mostly fixed in our pods (happy to be corrected if I'm wrong). I make a distinction between:
 * Fixed memory (**F**): % of memory that is consumed by a pod by their mere existence
 * Variable memory (**V**): % of memory consumed based on load.  

Let's call **L** the overall memory consumption required by the incoming load at a given point in time. That load is spread across the **N** pods available, so for each pod `V=L/N`

The memory usage in a pod will therefore be
`M = F + L / N`

What's the problem then? Using the scaling function:
`desiredReplicas=ceil[currentReplicas * ( currentMetricValue / desiredMetricValue )]`

Let's assume desiredMetricValue=90% and let's consider two scenarios for F=10% and F=80%. Let's display the resulting formula (play with graph [here](https://www.desmos.com/calculator/vpbioaratm))
`desiredReplicas=ceil[currentReplicas * ( F / 0.9 )]`

![image](https://github.com/mmeendez8/mmeendez8.github.io/assets/51260739/e72fb894-d982-4090-aff7-aaadd1c9cdf9)

What happens? When F is dominating, desiredReplicas is almost equal to currentReplicas, so it's not able to scale down. However, when V is dominating, that's when it is able to do so!

If there's a spike in traffic causing your memory to overload, that will trigger a new pod. But if that new pod's RAM is mostly bloated with stuff that is not related to traffic (deps or whatever), only a small % of its RAM will be used to alleviate the traffic load. 

When the traffic goes back to normal, the RAM in the pods will only go down slightly because it's mostly occupied with the fixed stuff. So the autoscaler will not be able to go back to 1 pod. Which is pointless, because you're monitoring memory unrelated to the traffic!